### DISCUSSION TITLE
Save trice COBS encoded data on target and view it later on PC

### DISCUSSION THREAD

**Discussion by tlsvc** [rokath/trice#253] [General]
Hi Thomas, 

I just stumbled upon this very interesting project via your article on memfault.com. Looks like really great work!

I'm wondering if trice could be used to save logs on the target "in field" and view those logs later on by using the PC tool. So is there an option in the PC tool to load the COBS data from a file? Or would I have to emulate a serial port?

Thanks in advance :)

Tristan

**Comment by rokath** [comment]
Hi Tristan, sorry for the delayed aswer. I just came accross your question. The anwer is yes:

`trice l -p FILE -args filename` 

This is possible in v0.48.0.

**Reply by tlsvc** [reply]
Thank you very much, that looks great!

**Comment by rokath** [comment]
For flash memory storage you could modify the encoding to reduce the data amount further.  You can drop the COBS  encoding and padding. Also no need for location information and maybe timestamps.

**Reply by rokath** [reply]
I opened an issue: https://github.com/rokath/trice/issues/260

**Reply by tlsvc** [reply]
Well, presumably I will use a SD-Card using FatFs to store the data, so flash memory space is not really a relevant constraint for me. I think it makes sense to just stick to the existing encoding and use it as a robust binary file format.

**Comment by rokath** [comment]
Anyway TREX and TCOBS are planned now for the future. The legacy encoding stays as option. See TriceV1.0 Spec draft in doc Folder.

**Comment by heezes** [comment]
Hi @rokath ,
I have been successfully using trice for saving device logs on memory and uploading it later for analysis. It works wonderfully.
Although i have one issue when using trice to decode the file is does not detect the EOF. Is it because of the software or there's some issue with the encoded file

**Comment by rokath** [comment]
The trice tool normally expects endless data. So, if an EOF is detected, it tries to read again a bit later. If you descibe your case in more detail, I could try to find a solution.

**Reply by heezes** [reply]
I'm using trice statements to log assertion and application code fault management. The trice data stream is stored in a file in Nand memory and is being uploaded to a remote server. Now what i usually do is download. decode(trice) and analyze the file. For automating the process i want the trice decoding to detect EOF so that I go ahead with analyze step, but due to the current behavior of trice software I'm stuck with the manual flow.

**Comment by rokath** [comment]
For JLink RTT, one use case is endless file reading. I could provide an additional CLI switch `-EOFStop`, but there are already a lot of switches. `-port FILE_UNTIL_END` instead of `-port FILE ` I like more but the name is too long. Or simply `-port EOF`? There is also the possibility to use the `-args` switch. But it carries the filename. What do you think?

**Reply by heezes** [reply]
I'm ok with any method which stops decoding as soon as EOF is detected. Although I think -args switch will provide more options to the user

**Comment by rokath** [comment]
You are right. We could take `:` as a delimiter between the filename and further optional switches inside the `-args` string. Would that be ok for your scripts? Only disadvantage: no file names with `:` in Linux. `\0` would be more universal but maybe cause other issues.

**Reply by heezes** [reply]
How about using a simple whitespace bw arguments or `-args "--name=file.bin --mode=EOF" could also be useful`

**Comment by rokath** [comment]
Hi @heezes, I am sorry for my lousy mind! Please try `-port FILEBUFFER` instead of using `-port FILE` with the *trice* version you are using right now. It should do what you want.

**Reply by heezes** [reply]
![image](https://user-images.githubusercontent.com/23493768/164155586-98438258-1e66-4844-bb7a-b18427eeed26.png)

**Reply by heezes** [reply]
Using `-port FILE` does decode the file.

**Reply by rokath** [reply]
You use slash in filename and backslash in til.json name. This should be ok. Please which trice version you are using? I cannot locate the error message in the current code. But I think the real problem is the `>` just before the trice in your command line. Please correct that and send me both complete command lines (FILE and FILE BUFFER)

**Reply by heezes** [reply]
![image](https://user-images.githubusercontent.com/23493768/164639790-3f53eab5-b70c-4fb2-96e5-abbde04750ce.png)
I'm using git bash in vscode and the os is windows 11.

**Reply by heezes** [reply]
![image](https://user-images.githubusercontent.com/23493768/164641746-dd14859d-e29c-46c6-98c6-4ed2c46c3007.png)
FILE BUFFER or FILEBUFFER is not present in the help menu although running the BUFFER command does not help as well

**Reply by rokath** [reply]
In the docs folder is TriceColor.md what may help to get color.  Concerning FILEBUFFER give me please a bit time.

**Reply by rokath** [reply]
Please check out latest release v0.49.1. I am sorry for the hazzle. FILEBUFFER was not existent in v0.48.0 and also not in the current master branch. I had to merge it from the trex branch. Please let me know if things work now.

**Reply by heezes** [reply]
I have downloaded the latest version but unfortunately things still are not working out. I have attached a til.json and raw file for you to check out as well. The trice does not exit after EOF.
NOTE: Change the trice.txt to trice.bin and til.txt to til.json as tbey were not attaching
[til.txt](https://github.com/rokath/trice/files/8570282/til.txt)
[trice.txt](https://github.com/rokath/trice/files/8570284/trice.txt)

**Reply by rokath** [reply]
Thanks for the files. I'll check in the next days.

**Reply by heezes** [reply]
@rokath  Any update on the issue?

**Reply by rokath** [reply]
Sorry, too busy right now. Hope next week to find time for this.

**Reply by heezes** [reply]
Cool. I wait for then.

**Comment by rokath** [comment]
PLease have a try with v0.49.2 and look at the closed issue #284.

**Reply by heezes** [reply]
Yeah its working. Thanks for the help!

**Comment by rokath** [comment]
:-)

**Comment by rokath** [comment]
With trice release v0.55.1 now TREX encoding with TCOBSv1 framing is default.  This means less bytes per trice.
